src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java

```diff
@@ -3808,45 +3807,15 @@ public boolean format(DateTimePrintContext context, StringBuilder buf) {
             if (inSecs == null) {
                 return false;
             }
-            long inSec = inSecs;
             int inNano = NANO_OF_SECOND.checkValidIntValue(inNanos != null ? inNanos : 0);
             // format mostly using LocalDateTime.toString
-            if (inSec >= -SECONDS_0000_TO_1970) {
-                // current era
-                long zeroSecs = inSec - SECONDS_PER_10000_YEARS + SECONDS_0000_TO_1970;
-                long hi = Math.floorDiv(zeroSecs, SECONDS_PER_10000_YEARS) + 1;
-                long lo = Math.floorMod(zeroSecs, SECONDS_PER_10000_YEARS);
-                LocalDateTime ldt = LocalDateTime.ofEpochSecond(lo - SECONDS_0000_TO_1970, 0, ZoneOffset.UTC);
-                if (hi > 0) {
-                    buf.append('+').append(hi);
-                }
-                buf.append(ldt);
-                if (ldt.getSecond() == 0) {
-                    buf.append(":00");
-                }
-            } else {
-                // before current era
-                long zeroSecs = inSec + SECONDS_0000_TO_1970;
-                long hi = zeroSecs / SECONDS_PER_10000_YEARS;
-                long lo = zeroSecs % SECONDS_PER_10000_YEARS;
-                LocalDateTime ldt = LocalDateTime.ofEpochSecond(lo - SECONDS_0000_TO_1970, 0, ZoneOffset.UTC);
-                int pos = buf.length();
-                buf.append(ldt);
-                if (ldt.getSecond() == 0) {
-                    buf.append(":00");
-                }
-                if (hi < 0) {
-                    if (ldt.getYear() == -10_000) {
-                        buf.replace(pos, pos + 2, Long.toString(hi - 1));
-                    } else if (lo == 0) {
-                        buf.insert(pos, hi);
-                    } else {
-                        buf.insert(pos + 1, Math.abs(hi));
-                    }
-                }
+            LocalDateTime ldt = LocalDateTime.ofEpochSecond(inSecs, fractionalDigits >= 0 ? 0 : inNano, ZoneOffset.UTC);
+            buf.append(ldt);
+            if (ldt.getSecond() == 0 && ldt.getNano() == 0) {
+                buf.append(":00");
             }
             // add fraction
-            if ((fractionalDigits < 0 && inNano > 0) || fractionalDigits > 0) {
+            if (fractionalDigits > 0) {
```

Comment:
This breaks the logic. `fractionalDigits` can be negative in the block below

Reply:
If fractionalDigits < 0, printNano is implemented in LocalDateTime

Reply:
`fractionalDigits == -2` is used to output 0, 3, 6 or 9 fractional digits as needed. This can be handled by `LocalDateTime`.

`fractionalDigits == -1`  is used to output as many fractional digits as needed, from 0 to 9 digits. This is NOT handled by `LocalDateTime`.

Furthermore, if the `-2` branch is handled by `LocalDateTime` then the logic below for `-2` is redundant.

If the tests did not fail as a result of this change, then I imagine the tests need improving.